tensorpipe/test/channel/channel_test.cc

```diff
@@ -15,513 +15,568 @@
-
-        // Initialize with zeroes.
-        std::vector<uint8_t> data(dataSize);
-        std::fill(data.begin(), data.end(), 0);
-
-        // Perform recv and wait for completion.
-        std::promise<Error> recvPromise;
-        std::mutex mutex;
-        std::unique_lock<std::mutex> callerLock(mutex);
-        auto descriptor = peers_->recv(PeerGroup::kClient);
-        channel->recv(
-            descriptor,
-            CpuTensor{data.data(), data.size()},
-            [&recvPromise, &mutex](const Error& error) {
-              std::unique_lock<std::mutex> calleeLock(mutex);
-              recvPromise.set_value(error);
-            });
-        callerLock.unlock();
-        Error recvError = recvPromise.get_future().get();
-        EXPECT_FALSE(recvError) << recvError.what();
-
-        // Validate contents of vector.
-        for (auto i = 0; i < data.size(); i++) {
-          EXPECT_EQ(data[i], i);
-        }
-
-        peers_->done(PeerGroup::kClient);
-        peers_->join(PeerGroup::kClient);
-
-        ctx->join();
-      });
-}
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> context =
+        this->helper_->makeContext("server");
+    this->peers_->send(PeerGroup::kClient, kReady);
+    context->createChannel(std::move(conn), Endpoint::kListen);
+  }
+
+  void client(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> context =
+        this->helper_->makeContext("client");
+    EXPECT_EQ(kReady, this->peers_->recv(PeerGroup::kClient));
+    context->createChannel(std::move(conn), Endpoint::kConnect);
+  }
+};
+
+CHANNEL_TEST_GENERIC(ContextIsNotJoined);
+
+// This test wants to make sure that the "heavy lifting" of copying data isn't
+// performed inline inside the recv method as that would make the user-facing
+// read method of the pipe blocking.
+// However, since we can't really check that behavior, we'll check a highly
+// correlated one: that the recv callback isn't called inline from within the
+// recv method. We do so by having that behavior cause a deadlock.
+class CallbacksAreDeferredTest : public ChannelTest<tensorpipe::CpuTensor> {
+  static constexpr auto dataSize = 256;
+
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<CpuContext> ctx = this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Initialize with sequential values.
+    std::vector<uint8_t> data(dataSize);
+    std::iota(data.begin(), data.end(), 0);
+
+    // Perform send and wait for completion.
+    std::promise<std::tuple<Error, TDescriptor>> descriptorPromise;
+    std::promise<Error> sendPromise;
+    std::mutex mutex;
+    std::unique_lock<std::mutex> callerLock(mutex);
+    channel->send(
+        CpuTensor{
+            .ptr = data.data(),
+            .length = data.size(),
+        },
```

Comment:
Goes back to a comment I left on an earlier diff: why use "keyword" args instead of "positional" args? It seems we're mixing the two styles.

Reply:
Will fix.